src/group_theory/sylow.lean

```diff
@@ -17,21 +17,141 @@ The Sylow theorems are the following results for every finite group `G` and ever
   `p`-subgroup in `G`.
 
-In this file, currently only the first of these results is proven.
+## Main definitions
+
+* `sylow p G` : The type of Sylow `p`-subgroups of `G`.
 
 ## Main statements
 
-* `exists_subgroup_card_pow_prime`: A generalisation of the first of the Sylow theorems: For every
-  prime power `pⁿ` dividing `G`, there exists a subgroup of `G` of order `pⁿ`.
+* `exists_subgroup_card_pow_prime`: A generalization of Sylow's first theorem:
+  For every prime power `pⁿ` dividing the cardinality of `G`,
+  there exists a subgroup of `G` of order `pⁿ`.
+* `is_p_group.exists_le_sylow`: A generalization of Sylow's first theorem:
+  Every `p`-subgroup is contained in a Sylow `p`-subgroup.
+* `sylow_conjugate`: A generalization of Sylow's second theorem:
+  If the number of Sylow `p`-subgroups is finite, then all Sylow `p`-subgroups are conjugate.
+* `card_sylow_modeq_one`: A generalization of Sylow's third theorem:
+  If the number of Sylow `p`-subgroups is finite, then it is congruent to `1` modulo `p`.
+-/
+
+open fintype mul_action subgroup
 
-## TODO
+section infinite_sylow
 
-* Prove the second and third of the Sylow theorems.
-* Sylow theorems for infinite groups
--/
+variables (p : ℕ) (G : Type*) [group G]
+
+/-- A Sylow `p`-subgroup is a maximal `p`-subgroup. -/
+structure sylow extends subgroup G :=
+(is_p_group' : is_p_group p to_subgroup)
+(is_maximal' : ∀ {Q : subgroup G}, is_p_group p Q → to_subgroup ≤ Q → Q = to_subgroup)
+
+variables {p} {G}
+
+instance : has_coe (sylow p G) (subgroup G) := ⟨sylow.to_subgroup⟩
+
+@[ext] lemma sylow.ext {P Q : sylow p G} (h : (P : subgroup G) = Q) : P = Q :=
+by cases P; cases Q; congr'
+
+lemma sylow.ext_iff {P Q : sylow p G} : P = Q ↔ (P : subgroup G) = Q :=
+⟨congr_arg coe, sylow.ext⟩
+
+/-- A generalization of **Sylow's first theorem**.
+  Every `p`-subgroup is contained in a Sylow `p`-subgroup. -/
+lemma is_p_group.exists_le_sylow {P : subgroup G} (hP : is_p_group p P) :
+  ∃ Q : sylow p G, P ≤ Q :=
+exists.elim (zorn.zorn_nonempty_partial_order₀ {Q : subgroup G | is_p_group p Q} (λ c hc1 hc2 Q hQ,
+⟨ { carrier := ⋃ (R : c), R,
+    one_mem' := ⟨Q, ⟨⟨Q, hQ⟩, rfl⟩, Q.one_mem⟩,
+    inv_mem' := λ g ⟨_, ⟨R, rfl⟩, hg⟩, ⟨R, ⟨R, rfl⟩, R.1.inv_mem hg⟩,
+    mul_mem' := λ g h ⟨_, ⟨R, rfl⟩, hg⟩ ⟨_, ⟨S, rfl⟩, hh⟩, (hc2.total_of_refl R.2 S.2).elim
+      (λ T, ⟨S, ⟨S, rfl⟩, S.1.mul_mem (T hg) hh⟩) (λ T, ⟨R, ⟨R, rfl⟩, R.1.mul_mem hg (T hh)⟩) },
+  λ ⟨g, _, ⟨S, rfl⟩, hg⟩, by
+  { refine exists_imp_exists (λ k hk, _) (hc1 S.2 ⟨g, hg⟩),
+    rwa [subtype.ext_iff, coe_pow] at hk ⊢ },
+  λ M hM g hg, ⟨M, ⟨⟨M, hM⟩, rfl⟩, hg⟩⟩) P hP) (λ Q ⟨hQ1, hQ2, hQ3⟩, ⟨⟨Q, hQ1, hQ3⟩, hQ2⟩)
+
+instance sylow.nonempty : nonempty (sylow p G) :=
+nonempty_of_exists is_p_group.of_bot.exists_le_sylow
+
+noncomputable instance sylow.inhabited : inhabited (sylow p G) :=
+classical.inhabited_of_nonempty sylow.nonempty
+
+instance sylow.mul_aut_mul_action : mul_action (mul_aut G) (sylow p G) :=
```

Comment:
Doesn't this generalize in the same way as `subgroup.pointwise_mul_action`?

Reply:
Are you suggesting that I add:
`instance sylow.mul_action' {α : Type*} [group α] [mul_distrib_mul_action α G] :
  mul_action α (sylow p G) :=
of_End_hom (to_End_hom.comp (mul_distrib_mul_action.to_mul_aut α G))`

Reply:
I'm suggesting you prove that instance instead of your one, via `..subgroup.pointwise_mul_action`

Reply:
I see. However, `is_p_group'` is a not a field of `mul_action`. Am I missing something?

Reply:
Whoops, I misspoke. I guess I mean something like:
```lean
{ smul := λ g S,
  { to_subgroup := g • S.to_subgroup,
    is_p_group' := sorry, },
  one_smul := sorry }
```

Reply:
Looks like it will work. I just need to reprove `sylow.smul_eq_iff_mem_normalizer`, because your approach changes the action from `comap` to `map` (it makes the action a little simpler actually, but `map` is a bit more painful to work with elementwise).

Reply:
I pushed a change. Is it along the lines of what you were thinking?